design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+
+* re-introduce `CREATE TABLE` statement to create RDBMS style tables that allow insert/update/delete statements
+* allow for complex queries over table state, e.g., aggregation and joins
+* we might not need to implement the new `AS OF <timestamp>` clause in the first iteration
+* we might not need to support terminating queries over streams right away
+
+We limit this KLIP to not introduce too much syntactic sugar to express queries.
+This may make some queries unnecessary "complex/clumsy" but the goal is to illustrate to underlying principles of the language.
+We can add syntactic sugar after we agree on the basic design of the language.
+
+
+
+## Value/Return
+
+The "Dual Query Language" should be easier to understand and easier to use.
+It will allow to support nested queries seamlessly and unifies query semantics (persistent vs. transient queries).
+Furthermore, the language is designed to be "future proof", i.e., it considers potential future features and thus allows
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
+* introduce `STREAM(<table>)` operator: takes a TABLE as input, and returns a STREAM that contains a record for each update to the TABLE
+  * note: the returned STREAM is a fact stream, and would be processed with corresponding STREAM semantics (it's semantically not a changelog stream)
+  * note: the input TABLE can either be an RDBMS style table of a MV
+  * by default, the output STREAM contains one record for each input table record (full table scan on the table generation when the query was issues)
+    plus a record for each input table update
+  * we could extend `STREAM()` to also accept a STREAM as input if it makes the language smoother;
+  it's obvious that `STREAM(<stream>)` would be idempotent, i.e., `STREAM(s) = s`
+* Introduce an `AS OF <timestamp>` clause for RDBMS style queries against tables
+* Introduce `CREATE MATERIALIZED VIEW` statement
+* Deprecate `CREATE TABLE` statement (in favor of the new `CREATE MATERIALIZED VIEW` statement) and eventually remove it
+* Deprecate `EMIT CHANGES` clause
+
+Semantic changes:
+
+* An aggregation query over a stream returns a stream (and not a table)
```

Comment:
Why this? The difference between tables and streams is the deduplication/update semantics. What do you propose here?

Reply:
I propose that the output stream contains a record for every update the "result table" (you can think of the result table as a `TEMP TABLE` in an RDBMS for this case, ie, the TABLE becomes an implementation detail).

The proposal is "necessary" to align to the overall design principles:
 - a stream query has an input stream and an output stream
 - it also allows for a straight forward support for "transient/client" queries. If the result would be a TABLE, a transient query would not return anything, even if a user rewrite the query to `STREAM(SELECT * FROM stream....)`

A more detailed discussion about this behavior is at the end of the KLIP: https://github.com/confluentinc/ksql/blob/888691db3d43e508ee1d87095288166117b501e2/design-proposals/klip-11-DQL.md#advanced-query-patterns

Hope this answers the question.